lib/auth/join_azure.go

```diff
@@ -0,0 +1,354 @@
+		}
+		cfg.verify = azureVerifyFuncFromOIDCVerifier(oidcConfig)
+	}
+
+	if cfg.certs == nil {
+		certs, err := getAzureCerts()
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		cfg.certs = certs
+	}
+	return nil
+}
+
+type azureRegisterOption func(cfg *azureRegisterConfig)
+
+// parseAndVeryAttestedData verifies that an attested data document was signed
+// by Azure. If verification is successful, it returns the ID of the VM that
+// produced the document.
+func parseAndVerifyAttestedData(adBytes []byte, challenge string, certs []*x509.Certificate) (string, error) {
+	var signedAD signedAttestedData
+	if err := utils.FastUnmarshal(adBytes, &signedAD); err != nil {
+		return "", trace.Wrap(err)
+	}
+	if signedAD.Encoding != "pkcs7" {
+		return "", trace.AccessDenied("unsupported signature type: %v", signedAD.Encoding)
+	}
+
+	sigPEM := "-----BEGIN PKCS7-----\n" + signedAD.Signature + "\n-----END PKCS7-----"
+	sigBER, _ := pem.Decode([]byte(sigPEM))
+	if sigBER == nil {
+		return "", trace.AccessDenied("unable to decode attested data document")
+	}
+
+	p7, err := pkcs7.Parse(sigBER.Bytes)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	fixAzureSigningAlgorithm(p7)
+
+	pool := x509.NewCertPool()
+	for _, cert := range certs {
+		pool.AddCert(cert)
+	}
+
+	if err := p7.VerifyWithChain(pool); err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	var ad attestedData
+	if err := utils.FastUnmarshal(p7.Content, &ad); err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	if ad.Nonce != challenge {
+		return "", trace.AccessDenied("challenge is missing or does not match")
+	}
+
+	return ad.ID, nil
+}
+
+// verifyVMIdentity verifies that the provided access token came from the
+// correct Azure VM.
+func verifyVMIdentity(ctx context.Context, cfg *azureRegisterConfig, accessToken, vmID string, requestStart time.Time) (*azure.VirtualMachine, error) {
+	tokenClaims, err := cfg.verify(ctx, accessToken)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	expectedIssuer := fmt.Sprintf("https://sts.windows.net/%v/", tokenClaims.TenantID)
+	// v2 tokens have the version appended to the issuer.
+	if tokenClaims.Version == "2.0" {
+		expectedIssuer += "2.0"
+	}
+
+	expectedClaims := jwt.Expected{
+		Issuer:   expectedIssuer,
+		Audience: jwt.Audience{azureAccessTokenAudience},
+		Time:     requestStart,
+	}
+
+	if err := tokenClaims.Validate(expectedClaims); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	rsID, err := arm.ParseResourceID(tokenClaims.ResourceID)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	vmClient := cfg.vmClient
+	if vmClient == nil {
+		tokenCredential := azure.NewStaticCredential(azcore.AccessToken{
+			Token:     accessToken,
+			ExpiresOn: tokenClaims.Expiry.Time(),
+		})
+		var err error
+		vmClient, err = azure.NewVirtualMachinesClient(rsID.SubscriptionID, tokenCredential, nil)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
+
+	vm, err := vmClient.Get(ctx, tokenClaims.ResourceID)
```

Comment:
Something just "clicked" for me when reading this - we are having the node export its access token to the teleport auth server, and the auth server can make requests with that token. From a security perspective this is pretty concerning. Especially for Teleport Cloud, it looks like we would be effectively exfiltrating customer access tokens, and those tokens would have the potential to be misused.

We do not know which managed identities may be attached to the node, it could be an identity with significant permissions within the customer account, which really should not leave the node/VM. I know the only necessary permission is `Microsoft.Compute/virtualMachines/read` but the azure admin may have attached anything.

Other join methods do not export access tokens like this - IAM join method only sends a pre-signed `GetCallerIdentity` request, OIDC join methods only send an ID token not an access token which can potentially make arbitrary requests.

Very sorry I did not see this earlier, but in my opinion it is not acceptable for us to export node access tokens in this way, what do you think?

Reply:
Is there any way we can limit the scope of the requested access token? If not, I'm not sure what other options we may have, besides documenting the behavior and maybe advising users to use limited identity specifically for joining purposes (and requiring to explicitly specify client ID), or going back to requiring them to run a separate service with Azure permissions.

AAD also has [ID tokens](https://learn.microsoft.com/en-us/azure/active-directory/develop/id-tokens) although it's not clear if it can be fetched for a VM/managed identity. Btw since we support OIDC joining, can it be used with Azure AD somehow? Even if it wouldn't allow to limit specifically by VM properties like resource group, etc. there may be other claims we could use? @atburke @nklaassen @strideynet @jakule

Reply:
Managed identities can have their scope limited, but I haven't found a way to request a limited scope for the tokens themselves, so that wouldn't help with a misconfigured identity.

Instance metadata doesn't appear to offer a way to fetch ID tokens, only access tokens.

Reply:
I think the issue is it may not be a misconfigured identity, but the vm may actually need some privileged identity for other purposes. If we just say they shouldn't use this join method for those nodes, and should only use it on nodes with this specific limited identity that's pretty limiting and confusing.

I think an alternative is to directly configure auth with the limited access token, rather than having nodes send their own token. The JWT could be hard coded in the provision token resource. I know we don't usually like to give customer cloud credentials to auth (for teleport cloud at least), but that's essentially what the current design does already, it's just that nodes send their own access token automatically and we don't save it anywhere (but a third party has no guarantee that we don't save it). In this idea the node would just send the attested data document, auth would look up the VM with the static access token to make sure it's in the right groups.

If there is no way to get an ID token (or similar) I think the only other option is a customer hosted discovery service

Reply:
So it sounds like we have 3 (at least) options:

1. We keep the current implementation where VMs send their access keys to Teleport during initial join and advise to use managed identity with limited permissions for IAM joining (in Azure, you can attach multiple identities to VMs) and have them specify identity ID explicitly in the token config.
2. We provide an option to set Azure API credentials (a token or client cert) in the provision token configuration but that would effectively mean storing a secret in Teleport.
3. We ask the customers who want to use Azure IAM joining deploy an agent (could be a part of discovery service) on their infrastructure with proper Azure API access and that agent will handle node verification. Compared to AWS IAM joining, they'd have to deploy this extra agent.

@klizhentas @xinding33 Do you have any preference of any of the option?

Reply:
I didn't realize you could attach multiple identities in azure and explicitly choose which one to send. It is a bit more configuration for the user but I think it could be acceptable.

Reply:
@klizhentas @xinding33 Thoughts?

Reply:
@atburke @jakule I chatted with @klizhentas about this today, we're ok with the current implementation. Let's just make sure to call this out in the guide we'll write and advise to use limited identity for joining.

Reply:
@r0mant in the current implementation (as this PR stands) a specific identity is not specified, there must be only a single identity attached to the VM and whatever that is will be used by default. Are you saying that is ok or should we force the admin to select a specific identity to use by name (and support VMs with multiple attached identities)?

Reply:
We should support multiple in any case. Let's have ability to specify one but use the default one by default.